### PR TITLE
Refactor how eth_requestAccounts works to make it more aligned with MetaMask

### PR DESCRIPTION
**Pull request by bbondy**
This work starts with a small refactor to pass error codes for provider instead of only a bool success code.
It then moves the provider errors into mojo.

If an account is selected and it is connected, eth_request and eth_accounts will return only that account to be consistent with MetaMask.

If an account is connected but the wallet is locked, we'll now ask to unlock the wallet instead of showing a permission prompt with the other accounts you can connect to.  Douglas is doing some UI changes to make it easier to switch which account is connected in a separate PR.

Finally we return request already in progress now instead of stacking requests. 

Resolves https://github.com/brave/brave-browser/issues/19750
Resolves https://github.com/brave/brave-browser/issues/19401
Resolves https://github.com/brave/brave-browser/issues/19641

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/brave_wallet/browser/brave_wallet_provider_impl.cc [hunk 1]
```diff
@@ -648,7 +665,12 @@ void BraveWalletProviderImpl::Locked() {
 }
 
 void BraveWalletProviderImpl::Unlocked() {
-  UpdateKnownAccounts();
+  if (pending_unlock_callback_) {
+    RequestEthereumPermissions(std::move(pending_unlock_callback_));
+    pending_unlock_callback_ = RequestEthereumPermissionsCallback();
```
Comment (yrliou) [inline components/brave_wallet/browser/brave_wallet_provider_impl.cc:670]: `pending_unlock_callback_` naming is confusing, I was a bit confused at first and didn't realize it's the callback for RequestEthereumPermissions, maybe pending_request_ethereum_permissions_callback_? 
And could you explain to me why do we need to construct a callback and assign to this member when the stored callback is used?

File: components/brave_wallet/browser/brave_wallet_provider_impl.cc [hunk 2]
```diff
@@ -542,6 +542,23 @@ void BraveWalletProviderImpl::OnRequestEthereumPermissions(
...
+          std::vector<std::string>(),
+          mojom::ProviderError::kUserRejectedRequest,
+          l10n_util::GetStringUTF8(IDS_WALLET_ALREADY_IN_PROGRESS_ERROR));
+      return;
+    }
+    pending_unlock_callback_ = std::move(callback);
+    keyring_controller_->RequestUnlock();
+    delegate_->ShowPanel();
+    return;
+  }
```
Comment (yrliou) [inline components/brave_wallet/browser/brave_wallet_provider_impl.cc:560]: Why do we need to run `RequestEthereumPermissions` once and handle the unlock logic here? Can this block be moved to `RequestEthereumPermissions`?

File: components/brave_wallet/common/brave_wallet.mojom [hunk 3]
```diff
@@ -402,6 +402,11 @@ interface KeyringController {
   // - At least 7 characters
   // - At least one non-alphanumeric character
   IsStrongPassword(string password) => (bool result);
+
+  // Obtains if there's a pending unlock request
+  // Note that there is no need for an API to notify unlocked
+  // because the KeyringControllerObserver Unlocked event can be used for that.
+  GetPendingUnlockRequest() => (bool pending);
```
Comment (yrliou) [inline components/brave_wallet/common/brave_wallet.mojom:409]: nit: How about `HasPendingUnlockRequest` or `IsUnlockRequestPending`?

File: components/brave_wallet/browser/brave_wallet_provider_impl.cc (hunk 1 above)
Comment (bbondy) [reply components/brave_wallet/browser/brave_wallet_provider_impl.cc]: Fixed in f466d97e0c79518206aab34f1e3793c5cf861930

File: components/brave_wallet/common/brave_wallet.mojom (hunk 3 above)
Comment (bbondy) [reply components/brave_wallet/common/brave_wallet.mojom]: Fixed in f466d97e0c79518206aab34f1e3793c5cf861930

File: components/brave_wallet/browser/brave_wallet_provider_impl.cc (hunk 2 above)
Comment (bbondy) [reply components/brave_wallet/browser/brave_wallet_provider_impl.cc]: I had that originally but there was some test failures that way. For example when no wallet is setup we want to open the onboarding.  Basically we only want to do this request unlock logic when things would be successful but the keyring is locked.

Comment (yrliou) [reply components/brave_wallet/browser/brave_wallet_provider_impl.cc]: I see, thanks for the explanation.

File: components/brave_wallet/browser/brave_wallet_provider_impl.cc [hunk 4]
```diff
@@ -608,7 +665,14 @@ void BraveWalletProviderImpl::Locked() {
 }
 
 void BraveWalletProviderImpl::Unlocked() {
-  UpdateKnownAccounts();
+  if (pending_request_ethereum_permissions_callback_) {
+    RequestEthereumPermissions(
+        std::move(pending_request_ethereum_permissions_callback_));
+    pending_request_ethereum_permissions_callback_ =
+        RequestEthereumPermissionsCallback();
```
Comment (yrliou) [inline components/brave_wallet/browser/brave_wallet_provider_impl.cc:672]: nit: I'll prefer this to be taken out if it turns out to be unnecessary.

**Review by yrliou — APPROVED** [review]

File: components/brave_wallet/browser/brave_wallet_provider_impl.cc (hunk 4 above)
Comment (bbondy) [reply components/brave_wallet/browser/brave_wallet_provider_impl.cc]: Removed in rebase

**Pull request merged**